src/middleware/validate.ts

```diff
@@ -0,0 +1,46 @@
+import { MiddlewareHandler } from "hono";
+import { ServiceType, Error } from "../types";
+import { toEnvKey } from "../utils";
+import { AppContext } from "../bindings";
+
+function validate(): MiddlewareHandler {
+  return async (context: AppContext, next) => {
+    const headers = context.req.raw.headers;
+    const xGatewayServiceHost = headers.get("x-gateway-service-host");
+    const xGatewayServiceType = (headers.get("x-gateway-service-type") as ServiceType) || "DIRECT";
+    const xGatewayServiceProxy = headers.get("x-gateway-service-proxy");
+    const xGatewayServiceToken = headers.get("x-gateway-service-token");
+
+    if (!xGatewayServiceHost) {
+      return context.json(<Error>{ error: "x-gateway-service-host header is required." }, 400);
+    }
+
+    if (xGatewayServiceType === "DIRECT" && xGatewayServiceProxy) {
+      return context.json(<Error>{ error: "x-gateway-service-proxy header is not allowed for direct service type." }, 400);
+    }
+
+    if (xGatewayServiceType === "GATEWAY" && !xGatewayServiceProxy) {
+      return context.json(<Error>{ error: "x-gateway-service-proxy header is required." }, 400);
+    }
+
+    // If the service token is not provided in the request headers, try to get it from the environment variables.
+    // The environment variable name is the service host name in uppercase with all non-alphanumeric characters replaced with "_".
+    const apiKey = toEnvKey(xGatewayServiceHost, xGatewayServiceProxy);
+    const token = xGatewayServiceToken || context.env[apiKey];
+
+    if (!token) {
+      return context.json(
+        <Error>{
+          error: "Cannot find API key for proxied service! Either provide it in the request headers or set it as an environment variable.",
+        },
+        400,
+      );
+    }
+
+    context.set("token", token);
```

Comment:
So I see we are using the context to set the token now. Is there a way to use an enum instead of having to use the "token" string itself?

Reply:
It is type-safe, ensuring you cannot misspell it. Is that something that's been bothering you?